File: day15/main.cpp
#include <iostream>
#include <utility>

static constexpr const std::pair A = { 512uLL, 16807u };
static constexpr const std::pair B = { 191uLL, 48271u };
static constexpr const auto mod = 2'147'483'647uLL;

template<typename T, typename U = int>
auto make_generator(T&& initializer, U&& divisible_by = 1)
{
    auto [value, mult] = initializer;
    
    auto next = [mult, divisible_by](auto&& val){
        do{ val = val * mult % mod; } while(val % divisible_by > 0);
        return val;
    };
 
    value = next(value);
    
    return [=] () mutable {
        return std::exchange(value, next(value));
    };
}

int main()
{
    {
        auto as = make_generator(A);
        auto bs = make_generator(B);
        int count{};
        for(int i = 0; i < 40'000'000; ++i)
            if((as() & 0xFFFF) == (bs() & 0xFFFF))
                ++count;
        std::cout << "Part 1: " << count << '\n';
    }
    {
        auto as = make_generator(A,4);
        auto bs = make_generator(B,8);
        int count{};
        for(int i = 0; i < 5'000'000; ++i)
            if((as() & 0xFFFF) == (bs() & 0xFFFF))
                ++count;
        std::cout << "Part 2: " << count << '\n';
    }
}

File: day15/main.rb
A = [512, 16807]
B = [191, 48271]


def make_generator(start, multiplier, multiples_of = 1)
    Enumerator.new do |enum|
        value = start
        while true
            value = value * multiplier % 2147483647
            enum.yield value & 0xFFFF if value % multiples_of == 0
        end
    end
end


P1 = make_generator(*A)
    .lazy
    .take(40_000_000)
    .zip(make_generator(*B))
    .count{ |l, r| l == r }


P2 = make_generator(*A, 4)
    .lazy
    .take(5_000_000)
    .zip(make_generator(*B, 8))
    .count{ |l, r| l == r }


puts "Part 1: %s" % P1
puts "Part 2: %s" % P2
